mount/mount.go

```diff
@@ -122,72 +227,40 @@ func (d *Datastore) Delete(key ds.Key) error {
 }
 
 func (d *Datastore) Query(q query.Query) (query.Results, error) {
-	if len(q.Filters) > 0 ||
-		len(q.Orders) > 0 ||
-		q.Limit > 0 ||
-		q.Offset > 0 {
-		// TODO this is still overly simplistic, but the only callers are
-		// `ipfs refs local` and ipfs-ds-convert.
-		return nil, errors.New("mount only supports listing all prefixed keys in random order")
-	}
 	prefix := ds.NewKey(q.Prefix)
 	dses, mounts, rests := d.lookupAll(prefix)
 
-	// current itorator state
-	var res query.Results
-	var mount ds.Key
-	i := 0
-
-	return query.ResultsFromIterator(q, query.Iterator{
-		Next: func() (query.Result, bool) {
-			var r query.Result
-			var more bool
-
-			for try := true; try; try = len(dses) > i {
-				if res == nil {
-					if len(dses) <= i {
-						//This should not happen normally
-						return query.Result{}, false
-					}
-
-					dst := dses[i]
-					mount = mounts[i]
-					rest := rests[i]
-
-					q2 := q
-					q2.Prefix = rest.String()
-					r, err := dst.Query(q2)
-					if err != nil {
-						return query.Result{Error: err}, false
-					}
-					res = r
-				}
+	// offset needs to be applied after the results are aggregated
+	offset := q.Offset
+	q.Offset = 0
 
-				r, more = res.NextSync()
-				if !more {
-					err := res.Close()
-					if err != nil {
-						return query.Result{Error: err}, false
-					}
-					res = nil
-
-					i++
-					more = len(dses) > i
-				} else {
-					break
-				}
-			}
+	queries := &querySet{
+		query: q,
+		heads: make([]*queryResults, 0, len(dses)),
+	}
 
-			r.Key = mount.Child(ds.RawKey(r.Key)).String()
-			return r, more
-		},
-		Close: func() error {
-			if len(mounts) > i && res != nil {
-				return res.Close()
-			}
-			return nil
-		},
-	}), nil
+	for i := range dses {
+		mount := mounts[i]
+		dstore := dses[i]
+		rest := rests[i]
+
+		qi := q
+		qi.Prefix = rest.String()
+		results, err := dstore.Query(qi)
+
+		if err != nil {
+			_ = queries.close()
+			return nil, err
+		}
+		queries.addResults(mount, results)
+	}
+
+	qr := query.ResultsFromIterator(q, query.Iterator{
+		Next:  queries.next,
+		Close: queries.close,
+	})
+
+	return query.NaiveLimit(query.NaiveOffset(qr, offset), q.Limit), nil
```

Comment:
This works, but I think we want limit/offset implemented in a manner similar to filters. Is that true?

Reply:
Maybe these things should only be applied if the `Limit` or `Offset` are non-zero?

Reply:
@michaelavila and I were just talking and we thought that maybe we should only wrap the results in NaiveLimit and NaiveOffset if those clauses exist in the Query to preserve the performance of existing code that obviously can't have ever used either of them.

Reply:
The code is actually updated to only apply limit/filter when limit or filter are greater than zero.